### PR TITLE
Fix ParseSSE function causing Leo to get cut off

### PR DESCRIPTION
**Pull request by nvonpentz**
Resolves https://github.com/brave/brave-browser/issues/34321

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by bbondy — APPROVED** [review]

**Comment by stephendonner** [comment]
## Verified `IN-PROGRESS` using both 

```
Brave | 1.62.67 Chromium: 119.0.6045.163 (Official Build) nightly (x86_64)
-- | --
Revision | 223789bf687abf95feb51f6c9fb2acdfd8c3a359
OS | macOS Version 14.2 (Build 23C5047e)
```

and

```
Brave | 1.62.67 Chromium: 119.0.6045.163 (Official Build) nightly (x86_64)
-- | --
Revision | 223789bf687abf95feb51f6c9fb2acdfd8c3a359
OS | macOS Version 14.2 (Build 23C5047e)
```

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win  11 x64` using the following build(s):

```
Brave | 1.62.67 Chromium: 119.0.6045.163 (Official Build) nightly (64-bit)
-- | --
Revision | 223789bf687abf95feb51f6c9fb2acdfd8c3a359
OS | Windows 11 Version 22H2 (Build 22621.2715)
```

Basically used the STR/guidance from https://github.com/brave/brave-core/pull/21018#issue-1997903655 and went through the following:

* ensured that you get a `Premium` prompt/modal within the sidebar panel when attempting to pick a premium model
* ensured that you can run through the signup flow and retrieve the needed tokens for users with a subscription

**`Example`** | **`Example`**
---------------|-----------------
Leo2 | Leo3

### `llama-2-13b-chat`

* ensured that you can ask the `llama` a question and get an answer that isn't cut off
* ensured that you can summarize a page/ask follow up questions without the answers being cut off

**`Example`** | **`Example`** | **`Example`** | **`Example`** | **`Example`**
----------------|----------------|----------------|----------------|------------------
image | image | image | image | image

### `llama-2-70b-chat`

* ensured that you can ask the `llama` a question and get an answer that isn't cut off
* ensured that you can summarize a page/ask follow up questions without the answers being cut off

**`Example`** | **`Example`** | **`Example`** | **`Example`** | **`Example`**
----------------|----------------|----------------|----------------|------------------
image | image | image | image | image

### `claude-instant-v1`

* ensured that you can ask the `claude` a question and get an answer that isn't cut off
* ensured that you can summarize a page/ask follow up questions without the answers being cut off

**`Example`** | **`Example`** | **`Example`** | **`Example`** | **`Example`**
----------------|----------------|----------------|----------------|------------------
image | image | image | image | image

File: components/api_request_helper/api_request_helper.cc [hunk 1]
```diff
@@ -453,7 +453,7 @@ void APIRequestHelper::URLLoaderHandler::ParseSSE(
         };
 
     DVLOG(2) << "Going to call ParseJson";
-    GetDataDecoder()->ParseJson(std::move(json.data()),
+    GetDataDecoder()->ParseJson(std::move(std::string(json)),
```
Comment (nullhook) [inline components/api_request_helper/api_request_helper.cc:456]: i'm curious how this behaved/s before and after. 

`std::move(json.data())` is unusual, doesn't make sense and behaved sporadically odd.

for a moment, let's examine this: `json.data()` returns a `const char*`, the consumer of this pointer reads char up to the null terminator string. 

how did this end up with sentence cut-offs?

File: components/api_request_helper/api_request_helper.cc [hunk 2]
```diff
@@ -453,7 +453,7 @@ void APIRequestHelper::URLLoaderHandler::ParseSSE(
         };
 
     DVLOG(2) << "Going to call ParseJson";
-    GetDataDecoder()->ParseJson(std::move(json.data()),
```
Comment (nullhook) [inline components/api_request_helper/api_request_helper.cc:456]: i'm curious how this behaved/s before and after.

`std::move(json.data())` is unusual, doesn't make sense and behaved sporadically odd.

for a moment, let's examine this: json.data() returns a `const char*`, the consumer of this pointer reads char up to the null terminator string.

how did this end up with sentence cut-offs?

Comment (bbondy) [reply components/api_request_helper/api_request_helper.cc:456]: https://bravesoftware.slack.com/archives/C05NDQP03AQ/p1700249757581809?thread_ts=1699393531.992589&cid=C05NDQP03AQ

**Pull request merged**